README.md
# BK Switch Theme JS
> Easily switch themes in your website

<img src="img/switcher.png" alt="gif" width="500" height="250" />

[![GitHub stars](https://img.shields.io/github/stars/PuneetGopinath/switch-theme-js)](https://github.com/PuneetGopinath/switch-theme-js/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/PuneetGopinath/switch-theme-js)](https://github.com/PuneetGopinath/switch-theme-js/network)
[![GitHub issues](https://img.shields.io/github/issues/PuneetGopinath/switch-theme-js)](https://github.com/PuneetGopinath/switch-theme-js/issues)
![GitHub repo size](https://img.shields.io/github/repo-size/PuneetGopinath/switch-theme-js)
![Lines of code](https://img.shields.io/tokei/lines/github/PuneetGopinath/switch-theme-js?label=total%20lines%20of%20code)
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.0-4baaaa.svg)](https://github.com/PuneetGopinath/switch-theme-js/blob/main/.github/CODE_OF_CONDUCT.md)
[![Join the chat at https://gitter.im/BaalKrshna/switch-theme](https://badges.gitter.im/BaalKrshna/switch-theme.svg)](https://gitter.im/BaalKrshna/switch-theme?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2FPuneetGopinath%2Fswitch-theme-js.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2FPuneetGopinath%2Fswitch-theme-js?ref=badge_shield)

## Features
- Themes switch without page reload
- Quick debugging
- Add your own themes and some themes ready-made in [`themes`](https://github.com/PuneetGopinath/switch-theme-js/tree/v1-dev/themes) folder
- Minified version of `index.js` automatically available in `index.min.js`
- And many more features coming

## v0
If you are looking for help with v0, you should head to the [`main` branch](https://github.com/PuneetGopinath/switch-theme-js/tree/main).
This branch is developing code for v1 release.

## Docs
See [`docs` folder](https://github.com/PuneetGopinath/switch-theme-js/tree/main/docs).

## Live demo
See [live demo](https://puneetgopinath.github.io/switch-theme-js/)

Click on "Dark theme" at the bottom, refresh the page, you see the page is default now set to dark theme.

## FAQs
These are the frequently asked questions:

### 1. What happens actually
This is a basic JS code, that stores the theme name in a cookie that stays for 10 days and adds the theme name to the `data-theme` attribute of the html tag.

### 2. How to design a theme?
You need to know CSS for that.

1. Think of a cool theme name
2. To select the theme in CSS, use `[data-theme="themeName"]` where `themeName` is your theme name. And design styles for the theme by selecting it.
3. To add a button for user to switch the theme (use the html `a` tag e.g. `<a href="javascript:SwitchThemeJS.switch('dark');">Dark theme</a>`)
4. Add `onload="SwitchThemeJS.get();"` to your `body` tag
5. Add `<script async defer src="https://cdn.jsdelivr.net/gh/PuneetGopinath/switch-theme-js@v1-dev/src/index.min.js"></script>` in the head tag.
6. Set a default theme by adding `data-theme` attribute to `html` tag, with your default theme name as value
7. That's all, test it out in your website, or submit a bug report for us to Analyse the bug.

## Contributing
You can find the contributing guidelines [here](https://github.com/PuneetGopinath/switch-theme-js/blob/main/.github/CONTRIBUTING.md)

## Changelog
See [CHANGELOG.md](https://github.com/PuneetGopinath/switch-theme-js/blob/main/CHANGELOG.md)

## Supporters
👏 A big thank you goes to all the supporters of this project!

You prove that the switch-theme-js Action has some value for you!

### Stargazers
[![Stargazers for @PuneetGopinath/switch-theme-js](https://reporoster.com/stars/PuneetGopinath/switch-theme-js)](https://github.com/PuneetGopinath/switch-theme-js/stargazers)

### Contributors
Thanks to all contributors again!

[![GitHub Contributors Image](https://contrib.rocks/image?repo=PuneetGopinath/switch-theme-js)](https://github.com/PuneetGopinath/switch-theme-js/contributors)

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2FPuneetGopinath%2Fswitch-theme-js.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2FPuneetGopinath%2Fswitch-theme-js?ref=badge_large)


CHANGELOG.md
# Changelog v1

The Changelog of the major version 1 of BK Switch Theme JS.

Note that the displayed date is in the format `dd-mm-yyyy`

[Older changelogs](#older-changelogs)

## [v1.0.0]
> **Released:** `TBA`

### Initial release!!

### Changes
- Renamed `src/switch.js` to `src/index.js`

### Breaking changes
- Now the functions are under a JS object i.e. `window.SwitchThemeJS`

### New features
- Easy debugging
- Minified version of `index.js` automatically available in `index.min.js`, minified using `UglifyJS`

### Docs
- Added `themes` folder

[v1.0.0]: https://github.com/PuneetGopinath/switch-theme-js/releases/tag/v1.0.0

## Older changelogs
- No older changelogs


src/switch.js
/**
 * Switch Theme JS v0.1.1
 * Copyright (c) 2021 The Baal Krshna Team
 */

function switchTheme(theme) {
    document.documentElement.dataset.theme = theme;
    setCookie("theme", theme, 10);
    console.log("Switched theme to: " + theme);
}

function getTheme() {
    var theme = getCookie("theme");
    if (theme !== "") {
        switchTheme(theme);
    }
    return theme;
}

function setCookie(name, value, exdays) {
    var d = new Date();
    d.setTime(d.getTime() + exdays * 24 * 60 * 60 * 1000);
    var expires = "expires=" + d.toUTCString();
    document.cookie = name + "=" + value + ";" + expires + ";path=/";
}

function getCookie(name) {
    var name = name + "=";
    var decodedCookie = decodeURIComponent(document.cookie);
    var ca = decodedCookie.split(";");
    for (var i = 0; i < ca.length; i++) {
        var c = ca[i];
        while (c.charAt(0) == " ") {
            c = c.substring(1);
        }
        if (c.indexOf(name) == 0) {
            return c.substring(name.length, c.length);
        }
    }
    return "";
}


src/index.js
/**
 * Switch Theme JS v1.0.0-alpha1
 * Copyright (c) 2021 The Baal Krshna Team and Contributors
 * @license Lesser General Public License v2.1 (LGPl-2.1 - https://opensource.org/licenses/lgpl-2.1.php)
 */

(function (root, ns, factory) {
    "use strict";
    root[ns] = factory(ns, root);
})(window, "SwitchThemeJS", function () {
    var setCookie = function (name, value, exdays) {
        var d = new Date();
        d.setTime(d.getTime() + exdays * 24 * 60 * 60 * 1000);
        var expires = "expires=" + d.toUTCString();
        document.cookie = name + "=" + value + ";" + expires + ";path=/";
    };
    var checkEmpty = function (variable) {
        if (variable) {
            return false;
        }
        return true;
    };
    var getCookie = function (cname) {
        const name = cname + "=";
        var decodedCookie = decodeURIComponent(document.cookie);
        var ca = decodedCookie.split(";");
        var i;
        var c;
        for (i = 0; i < ca.length; i++) {
            c = ca[i];
            while (c.charAt(0) == " ") {
                c = c.substring(1);
            }
            if (c.indexOf(name) == 0) {
                return c.substring(name.length, c.length);
            }
        }
        return "";
    };

    return {
        version: "v1.0.0-alpha1",
        get: function () {
            var theme = getCookie("theme");
            if (!checkEmpty(theme)) {
                this.switch(theme);
            } else {
                console.error("Theme not set or empty, can't get theme name");
            }
            return theme;
        },
        switch: function (theme) {
            document.documentElement.dataset.theme = theme;
            setCookie("theme", theme, 10);
            checkEmpty(theme) ? console.error("Theme not set or empty, can't switch theme") : console.log("Switched to: " + theme + " theme");
        },
    };
});
